core/modules/qdisp/QueryRequest.cc

```diff
@@ -213,6 +252,15 @@ bool QueryRequest::isCancelled() {
 }
 
 void QueryRequest::cleanup() {
+    LOGF_DEBUG("%1% QueryRequest::cleanup()" %_jobId);
+    {
+        std::lock_guard<std::mutex> lock(_finishStatusMutex);
+        if (_finishStatus == ACTIVE) {
+            LOGF_ERROR("%1% QueryRequest::cleanup called before finished" % _jobId);
+            return;
+        }
+    }
+    // These need to be outside the mutex lock, or you could delete the mutex before it is freed.
```

Comment:
I changed the comment to "These need to be outside the mutex lock, or you could delete _finishStatusMutex before it is unlocked." Resetting either one of _jobQuery or _keepAlive could result in this object being deleted immediately, or nothing happening to the object for a while. It depends on the state of the rest of the system. As long as either one is not null, then this object can't be deleted.